File: packages/sdk/src/scope.ts
/**
 * Query scope guard — block off-topic user prompts before they reach the LLM.
 * Zero LLM cost: keyword and regex rules only.
 */

export type ScopeMatchMode = "any_allowed" | "deny_only";

export interface TopicRuleInput {
  id?: string;
  name?: string;
  description?: string;
  keywords?: string[];
  terms?: string[];
  patterns?: string[];
}

export interface QueryScopeConfigInput {
  enabled?: boolean;
  purpose?: string;
  allowed_topics?: Array<TopicRuleInput | string>;
  denied_topics?: Array<TopicRuleInput | string>;
  deny_patterns?: string[];
  allow_greetings?: boolean;
  allow_empty?: boolean;
  match_mode?: ScopeMatchMode;
  refusal_message?: string;
}

export interface TopicRule {
  id: string;
  description: string;
  keywords: string[];
  patterns: string[];
}

export interface ScopeResult {
  allowed: boolean;
  reason: string;
  matchedTopic?: string;
  blockedBy?: string;
}

export interface QueryScopeConfig {
  enabled: boolean;
  purpose: string;
  allowedTopics: TopicRule[];
  deniedTopics: TopicRule[];
  denyPatterns: string[];
  allowGreetings: boolean;
  allowEmpty: boolean;
  matchMode: ScopeMatchMode;
  refusalMessage: string;
}

const DEFAULT_REFUSAL =
  "I can only help with topics related to my purpose. Please ask something within that scope.";

const GREETING_PATTERN =
  /^(hi|hello|hey|thanks|thank you|ok|okay|bye|goodbye|good morning|good evening)(\s|!|\?|\.|$)/i;

interface CompiledTopic {
  rule: TopicRule;
  keywordPatterns: RegExp[];
  regexPatterns: RegExp[];
}

interface CompiledDenyPattern {
  pattern: RegExp;
  label: string;
}

export class QueryScopeGuard {
  private readonly config: QueryScopeConfig;
  private readonly denyRegexes: CompiledDenyPattern[];
  private readonly allowed: CompiledTopic[];
  private readonly denied: CompiledTopic[];

  constructor(config: QueryScopeConfig) {
    this.config = config;
    this.denyRegexes = config.denyPatterns.map((p) => ({
      pattern: compilePattern(p),
      label: p,
    }));
    this.allowed = config.allowedTopics.map(compileTopicRule);
    this.denied = config.deniedTopics.map(compileTopicRule);
  }

  get enabled(): boolean {
    return this.config.enabled;
  }

  get purpose(): string {
    return this.config.purpose;
  }

  get refusalMessage(): string {
    return this.config.refusalMessage;
  }

  static disabled(): QueryScopeGuard {
    return new QueryScopeGuard({
      enabled: false,
      purpose: "",
      allowedTopics: [],
      deniedTopics: [],
      denyPatterns: [],
      allowGreetings: true,
      allowEmpty: true,
      matchMode: "any_allowed",
      refusalMessage: DEFAULT_REFUSAL,
    });
  }

  static fromConfig(input: QueryScopeConfigInput): QueryScopeGuard {
    const matchMode = input.match_mode ?? "any_allowed";
    if (matchMode !== "any_allowed" && matchMode !== "deny_only") {
      throw new Error("match_mode must be 'any_allowed' or 'deny_only'");
    }
    return new QueryScopeGuard({
      enabled: input.enabled ?? true,
      purpose: input.purpose ?? "",
      allowedTopics: (input.allowed_topics ?? []).map(parseTopicRule),
      deniedTopics: (input.denied_topics ?? []).map(parseTopicRule),
      denyPatterns: input.deny_patterns ?? [],
      allowGreetings: input.allow_greetings ?? true,
      allowEmpty: input.allow_empty ?? true,
      matchMode,
      refusalMessage: input.refusal_message ?? DEFAULT_REFUSAL,
    });
  }

  /** Load scope from JSON file (expects a top-level `scope` object or scope-only root). */
  static async loadJson(path: string): Promise<QueryScopeGuard> {
    const { readFile } = await import("node:fs/promises");
    const raw = JSON.parse(await readFile(path, "utf8")) as Record<string, unknown>;
    const scopeBlock = (raw.scope ?? raw) as QueryScopeConfigInput;
    if (typeof scopeBlock !== "object" || scopeBlock === null) {
      throw new Error(`Missing 'scope' object in ${path}`);
    }
    return QueryScopeGuard.fromConfig(scopeBlock);
  }

  check(query: string): ScopeResult {
    if (!this.config.enabled) {
      return { allowed: true, reason: "scope guard disabled" };
    }

    const text = query.trim();
    const normalized = text.toLowerCase();

    if (!text) {
      if (this.config.allowEmpty) {
        return { allowed: true, reason: "empty query allowed" };
      }
      return {
        allowed: false,
        reason: "empty query not allowed",
        blockedBy: "empty",
      };
    }

    if (this.config.allowGreetings && GREETING_PATTERN.test(text)) {
      return { allowed: true, reason: "greeting allowed", matchedTopic: "greeting" };
    }

    for (const { pattern, label } of this.denyRegexes) {
      if (pattern.test(text)) {
        return {
          allowed: false,
          reason: `matched deny pattern: ${label}`,
          blockedBy: label,
        };
      }
    }

    for (const compiled of this.denied) {
      if (topicMatches(compiled, normalized, text)) {
        return {
          allowed: false,
          reason: `matched denied topic: ${compiled.rule.id}`,
          blockedBy: compiled.rule.id,
        };
      }
    }

    if (this.config.matchMode === "deny_only") {
      return { allowed: true, reason: "deny-only mode, no deny match" };
    }

    if (this.allowed.length === 0) {
      return { allowed: true, reason: "no allowed_topics configured" };
    }

    for (const compiled of this.allowed) {
      if (topicMatches(compiled, normalized, text)) {
        return {
          allowed: true,
          reason: `matched allowed topic: ${compiled.rule.id}`,
          matchedTopic: compiled.rule.id,
        };
      }
    }

    return {
      allowed: false,
      reason: "query does not match any allowed topic",
      blockedBy: "out_of_scope",
    };
  }

  /** Return refusal text if denied, else `undefined` (safe to call LLM). */
  checkOrRefuse(query: string): string | undefined {
    const result = this.check(query);
    return result.allowed ? undefined : this.config.refusalMessage;
  }
}

function parseTopicRule(raw: TopicRuleInput | string): TopicRule {
  if (typeof raw === "string") {
    return { id: raw, description: "", keywords: [raw.toLowerCase()], patterns: [] };
  }
  const id = raw.id ?? raw.name;
  if (!id) {
    throw new Error("Topic rule requires 'id' or 'name'");
  }
  const keywords = (raw.keywords ?? raw.terms ?? []).map((k) => k.toLowerCase());
  return {
    id,
    description: raw.description ?? "",
    keywords,
    patterns: raw.patterns ?? [],
  };
}

function compilePattern(pattern: string): RegExp {
  try {
    return new RegExp(pattern, "i");
  } catch (err) {
    throw new Error(`Invalid deny pattern ${JSON.stringify(pattern)}: ${String(err)}`);
  }
}

function compileTopicRule(rule: TopicRule): CompiledTopic {
  const keywordPatterns = rule.keywords
    .filter((kw) => kw.trim())
    .map((kw) => new RegExp(escapeRegex(kw.toLowerCase()), "i"));
  const regexPatterns = rule.patterns.map(compilePattern);
  return { rule, keywordPatterns, regexPatterns };
}

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function topicMatches(compiled: CompiledTopic, normalized: string, original: string): boolean {
  if (compiled.keywordPatterns.some((p) => p.test(normalized))) return true;
  if (compiled.regexPatterns.some((p) => p.test(original))) return true;
  return false;
}

File: packages/sdk-go/scope.go
package acr

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

// ScopeMatchMode controls how allowed topics are evaluated.
type ScopeMatchMode string

const (
	ScopeMatchAnyAllowed ScopeMatchMode = "any_allowed"
	ScopeMatchDenyOnly   ScopeMatchMode = "deny_only"
)

const defaultScopeRefusal = "I can only help with topics related to my purpose. Please ask something within that scope."

var greetingPattern = regexp.MustCompile(`(?i)^(hi|hello|hey|thanks|thank you|ok|okay|bye|goodbye|good morning|good evening)(\s|!|\?|\.|$)`)

// TopicRule is one allowed or denied topic bucket.
type TopicRule struct {
	ID          string   `json:"id"`
	Description string   `json:"description,omitempty"`
	Keywords    []string `json:"keywords,omitempty"`
	Patterns    []string `json:"patterns,omitempty"`
}

// TopicRuleInput accepts YAML/JSON topic definitions (id or name, keywords or terms).
type TopicRuleInput struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Keywords    []string `json:"keywords"`
	Terms       []string `json:"terms"`
	Patterns    []string `json:"patterns"`
}

// QueryScopeConfigInput is the declarative scope configuration.
type QueryScopeConfigInput struct {
	Enabled         *bool            `json:"enabled"`
	Purpose         string           `json:"purpose"`
	AllowedTopics   []json.RawMessage `json:"allowed_topics"`
	DeniedTopics    []json.RawMessage `json:"denied_topics"`
	DenyPatterns    []string         `json:"deny_patterns"`
	AllowGreetings  *bool            `json:"allow_greetings"`
	AllowEmpty      *bool            `json:"allow_empty"`
	MatchMode       ScopeMatchMode   `json:"match_mode"`
	RefusalMessage  string           `json:"refusal_message"`
}

// QueryScopeConfig is the normalized scope configuration.
type QueryScopeConfig struct {
	Enabled        bool
	Purpose        string
	AllowedTopics  []TopicRule
	DeniedTopics   []TopicRule
	DenyPatterns   []string
	AllowGreetings bool
	AllowEmpty     bool
	MatchMode      ScopeMatchMode
	RefusalMessage string
}

// ScopeResult is the outcome of a scope check.
type ScopeResult struct {
	Allowed      bool
	Reason       string
	MatchedTopic string
	BlockedBy    string
}

type compiledTopic struct {
	rule            TopicRule
	keywordPatterns []*regexp.Regexp
	regexPatterns   []*regexp.Regexp
}

type compiledDenyPattern struct {
	pattern *regexp.Regexp
	label   string
}

// QueryScopeGuard filters user queries before they reach an LLM (zero LLM cost).
type QueryScopeGuard struct {
	config      QueryScopeConfig
	denyRegexes []compiledDenyPattern
	allowed     []compiledTopic
	denied      []compiledTopic
}

// DisabledScopeGuard returns a passthrough guard that always allows.
func DisabledScopeGuard() *QueryScopeGuard {
	return NewQueryScopeGuard(QueryScopeConfig{Enabled: false, RefusalMessage: defaultScopeRefusal})
}

// NewQueryScopeGuard builds a guard from normalized config.
func NewQueryScopeGuard(config QueryScopeConfig) *QueryScopeGuard {
	if config.RefusalMessage == "" {
		config.RefusalMessage = defaultScopeRefusal
	}
	if config.MatchMode == "" {
		config.MatchMode = ScopeMatchAnyAllowed
	}
	guard := &QueryScopeGuard{config: config}
	for _, p := range config.DenyPatterns {
		guard.denyRegexes = append(guard.denyRegexes, compiledDenyPattern{
			pattern: mustCompilePattern(p),
			label:   p,
		})
	}
	for _, rule := range config.AllowedTopics {
		guard.allowed = append(guard.allowed, compileTopicRule(rule))
	}
	for _, rule := range config.DeniedTopics {
		guard.denied = append(guard.denied, compileTopicRule(rule))
	}
	return guard
}

// FromScopeConfig parses declarative scope config (e.g. from JSON/YAML `scope` block).
func FromScopeConfig(input QueryScopeConfigInput) (*QueryScopeGuard, error) {
	enabled := true
	if input.Enabled != nil {
		enabled = *input.Enabled
	}
	allowGreetings := true
	if input.AllowGreetings != nil {
		allowGreetings = *input.AllowGreetings
	}
	allowEmpty := true
	if input.AllowEmpty != nil {
		allowEmpty = *input.AllowEmpty
	}
	matchMode := input.MatchMode
	if matchMode == "" {
		matchMode = ScopeMatchAnyAllowed
	}
	if matchMode != ScopeMatchAnyAllowed && matchMode != ScopeMatchDenyOnly {
		return nil, fmt.Errorf("match_mode must be 'any_allowed' or 'deny_only'")
	}

	allowed, err := parseTopicRules(input.AllowedTopics)
	if err != nil {
		return nil, err
	}
	denied, err := parseTopicRules(input.DeniedTopics)
	if err != nil {
		return nil, err
	}

	refusal := input.RefusalMessage
	if refusal == "" {
		refusal = defaultScopeRefusal
	}

	return NewQueryScopeGuard(QueryScopeConfig{
		Enabled:        enabled,
		Purpose:        input.Purpose,
		AllowedTopics:  allowed,
		DeniedTopics:   denied,
		DenyPatterns:   input.DenyPatterns,
		AllowGreetings: allowGreetings,
		AllowEmpty:     allowEmpty,
		MatchMode:      matchMode,
		RefusalMessage: refusal,
	}), nil
}

// LoadScopeJSON loads scope config from a JSON file (top-level `scope` or scope-only root).
func LoadScopeJSON(path string) (*QueryScopeGuard, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	block, ok := raw["scope"]
	if !ok {
		block = data
	}
	var input QueryScopeConfigInput
	if err := json.Unmarshal(block, &input); err != nil {
		return nil, err
	}
	return FromScopeConfig(input)
}

// Enabled reports whether scope checking is active.
func (g *QueryScopeGuard) Enabled() bool {
	return g.config.Enabled
}

// Purpose returns the documented agent purpose string.
func (g *QueryScopeGuard) Purpose() string {
	return g.config.Purpose
}

// RefusalMessage returns the user-facing refusal text.
func (g *QueryScopeGuard) RefusalMessage() string {
	return g.config.RefusalMessage
}

// Check returns whether the query may proceed to the LLM.
func (g *QueryScopeGuard) Check(query string) ScopeResult {
	if !g.config.Enabled {
		return ScopeResult{Allowed: true, Reason: "scope guard disabled"}
	}

	text := strings.TrimSpace(query)
	normalized := strings.ToLower(text)

	if text == "" {
		if g.config.AllowEmpty {
			return ScopeResult{Allowed: true, Reason: "empty query allowed"}
		}
		return ScopeResult{Allowed: false, Reason: "empty query not allowed", BlockedBy: "empty"}
	}

	if g.config.AllowGreetings && greetingPattern.MatchString(text) {
		return ScopeResult{Allowed: true, Reason: "greeting allowed", MatchedTopic: "greeting"}
	}

	for _, compiled := range g.denyRegexes {
		if compiled.pattern.MatchString(text) {
			return ScopeResult{
				Allowed:   false,
				Reason:    "matched deny pattern: " + compiled.label,
				BlockedBy: compiled.label,
			}
		}
	}

	for _, compiled := range g.denied {
		if topicMatches(compiled, normalized, text) {
			return ScopeResult{
				Allowed:   false,
				Reason:    "matched denied topic: " + compiled.rule.ID,
				BlockedBy: compiled.rule.ID,
			}
		}
	}

	if g.config.MatchMode == ScopeMatchDenyOnly {
		return ScopeResult{Allowed: true, Reason: "deny-only mode, no deny match"}
	}

	if len(g.allowed) == 0 {
		return ScopeResult{Allowed: true, Reason: "no allowed_topics configured"}
	}

	for _, compiled := range g.allowed {
		if topicMatches(compiled, normalized, text) {
			return ScopeResult{
				Allowed:      true,
				Reason:       "matched allowed topic: " + compiled.rule.ID,
				MatchedTopic: compiled.rule.ID,
			}
		}
	}

	return ScopeResult{
		Allowed:   false,
		Reason:    "query does not match any allowed topic",
		BlockedBy: "out_of_scope",
	}
}

// CheckOrRefuse returns refusal text if denied, else empty string (safe to call LLM).
func (g *QueryScopeGuard) CheckOrRefuse(query string) string {
	result := g.Check(query)
	if result.Allowed {
		return ""
	}
	return g.config.RefusalMessage
}

func parseTopicRules(raw []json.RawMessage) ([]TopicRule, error) {
	rules := make([]TopicRule, 0, len(raw))
	for _, item := range raw {
		var asString string
		if err := json.Unmarshal(item, &asString); err == nil {
			rules = append(rules, TopicRule{ID: asString, Keywords: []string{strings.ToLower(asString)}})
			continue
		}
		var input TopicRuleInput
		if err := json.Unmarshal(item, &input); err != nil {
			return nil, fmt.Errorf("invalid topic rule: %w", err)
		}
		id := input.ID
		if id == "" {
			id = input.Name
		}
		if id == "" {
			return nil, fmt.Errorf("topic rule requires 'id' or 'name'")
		}
		keywords := input.Keywords
		if len(keywords) == 0 {
			keywords = input.Terms
		}
		normalized := make([]string, 0, len(keywords))
		for _, kw := range keywords {
			normalized = append(normalized, strings.ToLower(kw))
		}
		rules = append(rules, TopicRule{
			ID:          id,
			Description: input.Description,
			Keywords:    normalized,
			Patterns:    input.Patterns,
		})
	}
	return rules, nil
}

func compileTopicRule(rule TopicRule) compiledTopic {
	keywords := make([]*regexp.Regexp, 0, len(rule.Keywords))
	for _, kw := range rule.Keywords {
		kw = strings.TrimSpace(kw)
		if kw == "" {
			continue
		}
		keywords = append(keywords, regexp.MustCompile("(?i)"+regexp.QuoteMeta(strings.ToLower(kw))))
	}
	regexPatterns := make([]*regexp.Regexp, 0, len(rule.Patterns))
	for _, p := range rule.Patterns {
		regexPatterns = append(regexPatterns, mustCompilePattern(p))
	}
	return compiledTopic{rule: rule, keywordPatterns: keywords, regexPatterns: regexPatterns}
}

func mustCompilePattern(pattern string) *regexp.Regexp {
	re, err := regexp.Compile("(?i)" + pattern)
	if err != nil {
		panic(fmt.Errorf("invalid pattern %q: %w", pattern, err))
	}
	return re
}

func topicMatches(compiled compiledTopic, normalized, original string) bool {
	for _, pattern := range compiled.keywordPatterns {
		if pattern.MatchString(normalized) {
			return true
		}
	}
	for _, pattern := range compiled.regexPatterns {
		if pattern.MatchString(original) {
			return true
		}
	}
	return false
}
